### PR TITLE
Nisyad/modify map per class

### PR DESCRIPTION
**Pull request by nisyad-ms**
This PR modifies the return output for `MeanAveragePrecision` for per class metrics. Return output as a dict with label_ids as keys.

### CODE REVIEW

**Comment by pj-ms** [comment]
pkg version update needed

File: visionmetrics/detection/mean_ap.py [hunk 1]
```diff
@@ -34,14 +34,19 @@ def __init__(self, box_format='xyxy', **kwargs):
...
     def update(self, predictions: List[List[List[float]]], targets: List[List[List[float]]]) -> None:
         predictions, targets = self._preprocess(predictions, targets)
+        self._update_label_ids(predictions, targets)
         super().update(predictions, targets)
 
     def compute(self):
         result = super().compute()
-        return {k: v for k, v in result.items() if k in ['map', 'map_50', 'map_75', 'map_per_class']}
+        if self.class_metrics:
+            result['map_per_class'] = {sorted(self._label_ids)[i]: v for i, v in enumerate(result['map_per_class'])}
```
Comment (pj-ms) [inline visionmetrics/detection/mean_ap.py:47]: this sort got run by len(result['map_per_class']) times

File: visionmetrics/detection/mean_ap.py [hunk 2]
```diff
@@ -64,3 +69,8 @@ def _convert_to_dict(boxes: List[List[float]], scores: bool = True) -> Dict[str,
             return {'boxes': boxes[:, -4:], 'labels': boxes[:, 0].to(torch.int), 'scores': boxes[:, 1].to(torch.float)}
         else:
             return {'boxes': boxes[:, -4:], 'labels': boxes[:, 0].to(torch.int)}
+
+    def _update_label_ids(self, predictions: List[Dict[str, torch.Tensor]], targets: List[Dict[str, torch.Tensor]]) -> None:
```
Comment (pj-ms) [inline visionmetrics/detection/mean_ap.py:73]: nit:

```
# for line 41
self._update_label_ids(predictions)
self._update_label_ids(targets)

def _update_label_ids(self, items: List[Dict[str, torch.Tensor]]]) -> None:
       for item in items:
            self._label_ids.update(item['labels'].tolist())
```

File: visionmetrics/detection/mean_ap.py [hunk 3]
```diff
@@ -34,14 +34,19 @@ def __init__(self, box_format='xyxy', **kwargs):
         if box_format != 'xyxy':
             raise ValueError(f'Expected box format to be "xyxy", got {box_format}')
         super().__init__(box_format=box_format, **kwargs)
+        self._label_ids = set()
```
Comment (pj-ms) [inline visionmetrics/detection/mean_ap.py:38]: maybe check the Len equality

File: visionmetrics/detection/mean_ap.py [hunk 4]
```diff
@@ -34,14 +34,19 @@ def __init__(self, box_format='xyxy', **kwargs):
...
```
Comment (pj-ms) [inline visionmetrics/detection/mean_ap.py:37]: There is reset and clone you might want to make sure they are working properly after adding this, with unit tests.

Comment (pj-ms) [reply visionmetrics/detection/mean_ap.py:37]: https://github.com/Lightning-AI/torchmetrics/blob/3b3b997487fea2e05a65d81abc23a7ebef3371f1/src/torchmetrics/metric.py#L658

File: tests/test_detection.py [hunk 5]
```diff
@@ -162,6 +162,35 @@ def test_cat_id_remap(self):
...
+        TARGETS = [[[0, 0, 0, 1, 1],
+                    [1, 0, 0, 0, 0]]]  # Wrong
+
+        metric = MeanAveragePrecision(iou_thresholds=[0.5], class_metrics=True)
+        metric.update(PREDICTIONS, TARGETS)
+        results = metric.compute()
+        self.assertAlmostEqual(results['map_per_class'].get(0), 1.0, places=5)
+        self.assertAlmostEqual(results['map_per_class'].get(1), 0.0, places=5)
+
+    def test_class_wise_with_missing_classes(self):
```
[the remark this answers is not in the document]
Comment (nisyad-ms) [reply tests/test_detection.py]: updated - separated into two tests

File: visionmetrics/detection/mean_ap.py (hunk 1 above)
Comment (nisyad-ms) [reply visionmetrics/detection/mean_ap.py]: thanks. updated

File: visionmetrics/detection/mean_ap.py (hunk 2 above)
Comment (nisyad-ms) [reply visionmetrics/detection/mean_ap.py]: updated

File: visionmetrics/detection/mean_ap.py (hunk 4 above)
Comment (nisyad-ms) [reply visionmetrics/detection/mean_ap.py:37]: added a unit test to verify label_ids set across batches.

Comment (pj-ms) [reply visionmetrics/detection/mean_ap.py:37]: **Clone test**
a = metric()
a.update(...., ....)
b = a.clone()
assert b.calculate() == a.calculate()

**Reset**
a = metric()
a.update(data)
metric_val = a.calculate()
a.reset()
a.update(data)
assert a.calculate() == metric_val

File: tests/test_detection.py [hunk 6]
```diff
@@ -162,6 +162,54 @@ def test_cat_id_remap(self):
         for k in result.keys():
             self.assertTrue(torch.allclose(result[k], result_remap_cat_id[k]))
 
+    def test_class_wise_metrics(self):
+        PREDICTIONS = [[[0, 1.0, 0, 0, 1, 1],
+                        [1, 1.0, 0, 0, 1, 1]]]
+
+        TARGETS = [[[0, 0, 0, 1, 1],
+                    [1, 0, 0, 0, 0]]]  # Wrong
```
Comment (pj-ms) [inline tests/test_detection.py:170]: slightly confusing

**Review by shonohs — APPROVED** [review]

File: visionmetrics/detection/mean_ap.py (hunk 4 above)
Comment (nisyad-ms) [reply visionmetrics/detection/mean_ap.py]: Updated. 

Now using existing *-labels attributes of the Metric class to track all 'seen' labels. Added unit tests.

File: tests/test_detection.py (hunk 6 above)
Comment (nisyad-ms) [reply tests/test_detection.py]: Fixed and updated.

File: tests/test_detection.py [hunk 7]
```diff
@@ -162,6 +161,77 @@ def test_cat_id_remap(self):
...
+
+        metric2 = MeanAveragePrecision(iou_thresholds=[0.5], class_metrics=True)
+        metric2.update(PREDICTIONS, TARGETS)
+        results2 = metric2.compute()
+
+        self.assertEqual(results['map_per_class'].get(1, "Does not exist"), "Does not exist")
+        self.assertAlmostEqual(results['map_per_class'].get(2), 1.0, places=5)
+        self.assertEqual(results, results2)
+
+    def test_class_wise_after_clone(self):
```
Comment (pj-ms) [inline tests/test_detection.py:221]: where is the clone tested?

File: tests/test_detection.py [hunk 8]
```diff
@@ -162,6 +161,77 @@ def test_cat_id_remap(self):
...
+        TARGETS = [[[2, 0, 0, 1, 1]]]
+
+        metric = MeanAveragePrecision(iou_thresholds=[0.5], class_metrics=True)
+        metric.update(PREDICTIONS, TARGETS)
+        results = metric.compute()
+        self.assertAlmostEqual(results['map_per_class'].get(1), -1.0, places=5)
+        self.assertAlmostEqual(results['map_per_class'].get(2), 1.0, places=5)
+
+    def test_class_wise_after_reset(self):
+        PREDICTIONS = [[[1, 1.0, 0, 0, 1, 1]]]
```
Comment (pj-ms) [inline tests/test_detection.py:200]: Can you have multiple test cases covering different scenarios:

more classes before reset
more classes after reset
..

Wrt info on parametrized tests:
https://chat.openai.com/share/9841c3a6-3be4-4450-b3a5-b6d7cbe81250

File: tests/test_detection.py (hunk 7 above)
Comment (pj-ms) [reply tests/test_detection.py:221]: More test cases pls. Same with reset

File: visionmetrics/detection/mean_ap.py [hunk 9]
```diff
@@ -37,11 +37,15 @@ def __init__(self, box_format='xyxy', **kwargs):
...
     def update(self, predictions: List[List[List[float]]], targets: List[List[List[float]]]) -> None:
         predictions, targets = self._preprocess(predictions, targets)
+
         super().update(predictions, targets)
 
     def compute(self):
         result = super().compute()
-        return {k: v for k, v in result.items() if k in ['map', 'map_50', 'map_75', 'map_per_class', 'classes']}
+        if self.class_metrics:
+            result['map_per_class'] = {k.tolist(): v for k, v in zip(result['classes'], result['map_per_class'])}
```
Comment (pj-ms) [inline visionmetrics/detection/mean_ap.py:46]: Any reason still keeping this behavior: 
- the base class returns list, while this class returns dictionary for result['map_per_class']?

Comment (pj-ms) [reply visionmetrics/detection/mean_ap.py:46]: I remember in our last discussion, we agreed the need of dictionary output is mostly one kind of client need.

It does have the benefit wrt clarity, but there are ways doing it without making it a dictionary making it behave differently than the base class.

If no code change is needed, you can still check in the tests which could be helpful

**Pull request closed without merging**